src/js/game/LevelMVC/LevelPlane.js

```diff
@@ -325,10 +326,168 @@ module.exports = class LevelPlane extends Array {
+        this.activatePiston(this.indexToCoordinates(index));
+        if (this.levelModel) {
+          this.levelModel.controller.updateFowPlane();
+          this.levelModel.controller.updateShadingPlane();
+        }
+      } else if (!this[index].isPowered) {
+        this.deactivatePiston(this.indexToCoordinates(index));
+        if (this.levelModel) {
+          this.levelModel.controller.updateFowPlane();
+          this.levelModel.controller.updateShadingPlane();
+        }
+      }
+    }
+  }
+
+  /**
+  * Activates a piston at a given position to push blocks away from it depending on type.
+  */
+  activatePiston(position) {
+    let neighbors = this.getOrthogonalBlocks(position);
+    let neighborPosition = this.getOrthogonalPositions(position);
+
+    switch (this[this.coordinatesToIndex(position)].blockType) {
+      case "pistonUp": {
+        if (neighbors.north.block.blockType !== "" && neighbors.north.block.blockType !== "pistonArm" ) {
+          let offset = [0,-1];
+          let blocksPositions = this.getBlocksToPush(neighborPosition[0], offset[0], offset[1]);
+          this.pushBlocks(blocksPositions, offset[0], offset[1]);
+        } else if (neighbors.north.block.blockType === "") {
+          let armBlock = new LevelBlock("pistonArm");
+          this.setBlockAt(neighborPosition[0], armBlock);
+        }
+        break;
+      }
+      case "pistonDown": {
+        if (neighbors.south.block.blockType !== "" && neighbors.south.block.blockType !== "pistonArm") {
+          let offset = [0,1];
+          let blocksPositions = this.getBlocksToPush(neighborPosition[1], offset[0], offset[1]);
+          this.pushBlocks(blocksPositions, offset[0], offset[1]);
+        } else if (neighbors.south.block.blockType === "") {
+          let armBlock = new LevelBlock("pistonArm");
+          this.setBlockAt(neighborPosition[1], armBlock);
+        }
+        break;
+      }
+      case "pistonRight": {
+        if (neighbors.east.block.blockType !== "" && neighbors.east.block.blockType !== "pistonArm") {
+          let offset = [1,0];
+          let blocksPositions = this.getBlocksToPush(neighborPosition[2], offset[0], offset[1]);
+          this.pushBlocks(blocksPositions, offset[0], offset[1]);
+        } else if (neighbors.east.block.blockType === "") {
+          let armBlock = new LevelBlock("pistonArm");
+          this.setBlockAt(neighborPosition[2], armBlock);
+        }
+        break;
+      }
+      case "pistonLeft": {
+        if (neighbors.west.block.blockType !== "" && neighbors.west.block.blockType !== "pistonArm") {
+          let offset = [-1,0];
+          let blocksPositions = this.getBlocksToPush(neighborPosition[3], offset[0], offset[1]);
+          this.pushBlocks(blocksPositions, offset[0], offset[1]);
+        } else if (neighbors.west.block.blockType === "") {
+          let armBlock = new LevelBlock("pistonArm");
+          this.setBlockAt(neighborPosition[3], armBlock);
+        }
```

Comment:
This is a lot of repeated code; I think it would be a lot nicer to just assign `neighbor`, `position`, and `offset` in the case block and then just do

        if (neighbor.block.blockType !== "" && neighbor.block.blockType !== "pistonArm" ) {
          let blocksPositions = this.getBlocksToPush(neighborPosition[position], offset[0], offset[1]);
          this.pushBlocks(blocksPositions, offset[0], offset[1]);
        } else if (neighbor.block.blockType === "") {
          let armBlock = new LevelBlock("pistonArm");
          this.setBlockAt(neighborPosition[position], armBlock);
        }

once afterwards

Reply:
👍